signature/src/balrog.rs

```diff
@@ -0,0 +1,363 @@
+            let _ = Self::verify_cert_chain_pair(subject, issuer, current_time)?;
+            index = index + 1;
+        }
+
+        /* Verify the root certificate. */
+        let root_cert = match self.chain.last() {
+            None => return Err(BalrogError::CertificateNotFound),
+            Some(x) => x,
+        };
+        let _ = Self::verify_cert_chain_pair(root_cert, root_cert, current_time)?;
+
+        let hash_stripped = root_hash.replace(":", "");
+        let hash_decoded = match hex::decode(hash_stripped) {
+            Err(_e) => return Err(BalrogError::RootHashParseFailed),
+            Ok(x) => x,
+        };
+        if self.root_hash.is_empty() || (self.root_hash != hash_decoded) {
+            return Err(BalrogError::RootHashMismatch);
+        }
+
+        /* Success! */
+        Ok(())
+    }
+
+    /* Internal helper function to check that a certificate was signed by its issuer. */
+    fn verify_cert_chain_pair(
+        subject: &X509Certificate,
+        issuer: &X509Certificate,
+        current_time: i64,
+    ) -> Result<(), BalrogError> {
+        /* Check that the subject and issuer names match. */
+        if subject.issuer() != issuer.subject() {
+            return Err(BalrogError::ChainSubjectMismatch);
+        }
+
+        /* Check the certificate validity period. */
+        let asn_timestamp = ASN1Time::from_timestamp(current_time)?;
+        if !subject.validity().is_valid_at(asn_timestamp) {
+            return Err(BalrogError::CertificateExpired);
+        }
+        if !issuer.validity().is_valid_at(asn_timestamp) {
+            return Err(BalrogError::CertificateExpired);
+        }
+
+        /* Check that the issuer is permitted to sign certificates. */
+        let ca_key_usage = match issuer.key_usage()? {
+            None => return Err(BalrogError::IssuerUnauthorized),
+            Some(x) => x.value,
+        };
+        if !ca_key_usage.key_cert_sign() {
+            return Err(BalrogError::IssuerUnauthorized);
+        }
+        if !issuer.is_ca() {
+            return Err(BalrogError::IssuerUnauthorized);
+        }
+
+        /* Check X509 certificate validation. */
+        let mut logger = VecLogger::default();
+        // structure validation status
+        let ok = X509StructureValidator
+            .chain(X509CertificateValidator)
+            .validate(subject, &mut logger);
+        if !ok {
+            let reason = match logger.errors().first() {
+                None => "Unknown error",
+                Some(s) => s,
+            };
+            return Err(BalrogError::X509ValidationError(reason.to_string()));
+        }
+
+        /* Verify the cryptographic signature. */
+        Ok(subject.verify_signature(Some(issuer.public_key()))?)
+    }
+
+    /* Check a hostname against the leaf certificate. */
+    pub fn verify_leaf_hostname(&self, hostname: &str) -> Result<(), BalrogError> {
+        let leaf = match self.chain.first() {
```

Comment:
For my understanding: What makes a certificate a leaf certificate? Are all non-root certificate considered leaf certificates?

Reply:
The terms as I have been using them are:
- `leaf`: the very first certificate in the chain, it's the one that actually signed the Balrog update data that the client gets.
- `root`: the very last certificate in the chain, it's the trust anchor for the whole update system.
- `intermediate`: any certificate in the middle of the chain.

Reply:
Got it, I asked because I read in other places on the internet(tm) "leaf" being used to refer to all other certificates that are not the root. Would you mind adding a comment with your terminology? (Unless it's super obvious, in that case let's pretend I never asked).